common/ferrostar/src/navigation_controller/mod.rs

```diff
@@ -431,15 +451,22 @@ mod tests {
                 TripState::Navigating {
                     current_step_geometry_index,
                     ref remaining_steps,
+                    ref deviation,
                     ..
                 } => {
                     if let Some(index) = current_step_geometry_index {
                         let geom_length = remaining_steps[0].geometry.len() as u64;
+                        // Regression test that the geometry index is valid
                         assert!(
                             index < geom_length,
                             "index = {index}, geom_length = {geom_length}"
                         );
                     }
+
+                    // Regression test that we are never marked as off the route.
+                    // We used to encounter this with relative step advance on self-intersecting
+                    // routes, for example.
+                    assert_eq!(deviation, &RouteDeviation::NoDeviation);
```

Comment:
In the case of the self-intersecting route, an improper step advance would eventually result in a route deviation (which is why our test case eventually reroutes).